### PR TITLE
feat: id stitching improvement in shopify pixel transformations

### PR DESCRIPTION
**Pull request by yashasvibajpai**
## What are the changes introduced in this PR?

- We are handling `userId` enrichment from redis for anonymous as well as logged in users checkout scenarios.
We will be getting the value of userId as customerId that was sent in the rudderIdentifier event and stored in redis as anonymousId <-> userId mapping

- We will also multiplex the ecom events where anonymousId and userId both are present, into an identify call also so as to allow downstream identification of these profiles properly.

Other refactors include fixing some mappings that were being referenced from v0 source code have fixed those.

## What is the related Linear task?

Resolves INT-3009

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/334/index.html

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/361/index.html

File: src/v1/sources/shopify/webpixelTransformations/pixelTransform.js [hunk 1]
```diff
@@ -161,12 +161,19 @@ function processPixelEvent(inputEvent) {
     message.context.campaign = campaignParams;
   }
   message.messageId = id;
+
+  // attach userId to the message if anonymousId is present in Redis
+  // this allows stitching of events from the same user across multiple checkouts
+  const redisData = await RedisDB.getVal(`pixel:${message.anonymousId}`);
+  if (isDefinedNotNullNotEmpty(redisData)) {
```
Comment (krishna2020) [inline src/v1/sources/shopify/webpixelTransformations/pixelTransform.js:168]: Also have condition to check message.userId is not null

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/378/index.html

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/384/index.html

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/396/index.html

File: src/v1/sources/shopify/webpixelTransformations/pixelTransform.js [hunk 2]
```diff
@@ -116,15 +113,13 @@ function processPixelEvent(inputEvent) {
       break;
     case PIXEL_EVENT_TOPICS.CHECKOUT_STARTED:
     case PIXEL_EVENT_TOPICS.CHECKOUT_COMPLETED:
-      if (customer.id) message.userId = customer.id || '';
```
Comment (yashasvibajpai) [inline src/v1/sources/shopify/webpixelTransformations/pixelTransform.js:119]: We don't ever get this id in the pixel events in both anonymous or logged in users events. So it is better to remove it, and keep the userId assignment pointed to redis, thus ensuring a single source of truth.

File: src/v1/sources/shopify/utils.js [hunk 3]
```diff
@@ -8,11 +8,23 @@ const NO_OPERATION_SUCCESS = {
   statusCode: 200,
 };
 
+/**
+ * Updates the anonymousId to userId mapping in Redis
+ * @param {String} anonymousId
+ * @param {String} userId
+ */
+const updateAnonymousIdToUserIdInRedis = async (anonymousId, userId) => {
```
Comment (yashasvibajpai) [inline src/v1/sources/shopify/utils.js:16]: moved this to outside util as the userId setting in redis done here is used by both pixel and server side events.

File: src/v1/sources/shopify/utils.test.js [hunk 4]
```diff
@@ -45,4 +49,22 @@ describe('Identifier Utils Tests', () => {
       await expect(processIdentifierEvent(event)).rejects.toThrow('Redis connection failed');
     });
   });
+
+  describe('test updateAnonymousIdToUserIdInRedis', () => {
+    beforeEach(() => {
+      jest.clearAllMocks();
+    });
+
+    it('should update the anonymousId to userId in redis', async () => {
```
Comment (krishna2020) [inline src/v1/sources/shopify/utils.test.js:58]: Add negative test case with null/undefined userId

File: src/v1/sources/shopify/transformV1.js [hunk 5]
```diff
@@ -17,7 +17,7 @@ const processV1Events = async (event) => {
       return processIdentifierEvent(event);
     }
     // handle events from the app pixel.
-    const pixelWebEventResponse = processPixelWebEvents(event);
+    const pixelWebEventResponse = await processPixelWebEvents(event);
```
Comment (krishna2020) [inline src/v1/sources/shopify/transformV1.js:20]: what got changes to add `await` ?

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js [hunk 6]
```diff
@@ -33,7 +35,7 @@ const NO_OPERATION_SUCCESS = {
 const identifyPayloadBuilder = (event) => {
   const message = new Message(INTEGERATION);
   message.setEventType(EventType.IDENTIFY);
-  message.setPropertiesV2(event, MAPPING_CATEGORIES[EventType.IDENTIFY]);
+  message.setPropertiesV2(event, identifyMappingJSON);
```
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:38]: I don't see any difference in the two `identify` mapping configuration

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:1]: ecomPayloadBuilder has duplicate code  for event.customer mappings

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js [hunk 7]
```diff
@@ -111,11 +113,16 @@ const processEvent = async (inputEvent, metricMetadata) => {
   // attach anonymousId if the event is track event using note_attributes
   if (message.type !== EventType.IDENTIFY) {
     await setAnonymousId(message, event, metricMetadata);
+    await updateAnonymousIdToUserIdInRedis(message);
```
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:116]: message signature for `updateAnonymousIdToUserIdInRedis` is wrong

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js [hunk 8]
```diff
@@ -111,11 +113,16 @@ const processEvent = async (inputEvent, metricMetadata) => {
...
   }
   // attach userId, email and other contextual properties
   message = handleCommonProperties(message, event, shopifyTopic);
   // add cart_token_hash to traits if cart_token is present
   message = addCartTokenHashToTraits(message, event);
+  const redisData = await RedisDB.getVal(`pixel:${message.anonymousId}`);
```
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:122]: remove `userId` handling from `handleCommonProperties` & keep it at this level along with logic of enrichment from redis data

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js [hunk 9]
```diff
@@ -111,11 +113,16 @@ const processEvent = async (inputEvent, metricMetadata) => {
   // attach anonymousId if the event is track event using note_attributes
   if (message.type !== EventType.IDENTIFY) {
```
Comment (krishna2020) [inline src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:114]: do you need any of post transformation steps for identify event ?
```suggestion
  if (message.type == EventType.IDENTIFY) return message;
```

File: src/v1/sources/shopify/webpixelTransformations/pixelTransform.js (hunk 2 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webpixelTransformations/pixelTransform.js:119]: No, always seen null values populate there even if the user is logged in.

File: src/v1/sources/shopify/transformV1.js (hunk 5 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/transformV1.js:20]: `const redisData = await RedisDB.getVal(`pixel:${message.anonymousId}`);`
The user id enrichment call to redis is an asynchronous operation. The parent functions need to be marked async to use await and properly handle the Promise chain.
Any function that calls `processPixelWebEvents` would also need to be async and use await if it needs to wait for the pixel event processing to complete.

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js (hunk 6 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:38]: Yes, just wanted there should be least dependency on old v0 source code, so as to allow granular control w.r.t new v1 source transformations without any concerns for supporting backwards compatibility.

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:1]: removed one of the redundant mapping and also now using the identify mapping from v1

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js (hunk 9 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:114]: yes, handling common properties (like adding topic, shopify details and warehouse-related json paths) as well as after the redis setting for userId, we use that userId in all events whether identify or track

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js (hunk 7 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js]: thanks, updated

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/415/index.html

File: src/v1/sources/shopify/webhookTransformations/serverSideTransform.js (hunk 9 above)
Comment (krishna2020) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:110]: And why not trigger `updateAnonymousIdToUserIdInRedis` in case of identify as we extended the signature of identify call to capture userId too now

Comment (krishna2020) [reply src/v1/sources/shopify/webhookTransformations/serverSideTransform.js:110]: identify events do not have cartToken, so getting anonId is not possible

**Review by krishna2020 — APPROVED** [review]

File: src/v1/sources/shopify/utils.js [hunk 10]
```diff
@@ -8,11 +8,27 @@ const NO_OPERATION_SUCCESS = {
...
 const isIdentifierEvent = (payload) => ['rudderIdentifier'].includes(payload?.event);
 
 const processIdentifierEvent = async (event) => {
-  const { cartToken, anonymousId } = event;
-  await RedisDB.setVal(`pixel:${cartToken}`, ['anonymousId', anonymousId]);
+  const { cartToken, anonymousId, userId } = event;
+  if (cartToken && anonymousId) {
+    // set the cartToken to anonymousId mapping in Redis for 12 hours
+    await RedisDB.setVal(`pixel:${cartToken}`, ['anonymousId', anonymousId], 43200);
+  }
```
Comment (utsabc) [inline src/v1/sources/shopify/utils.js:30]: Make changes based on `action` field of the identifier event

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/439/index.html

File: src/v1/sources/shopify/utils.js [hunk 11]
```diff
@@ -22,13 +22,20 @@ const updateAnonymousIdToUserIdInRedis = async (anonymousId, userId) => {
...
+/**
+ * Sets the cartToken <-> anonymousId mapping or anonymousId <-> userId mapping in Redis based on the event action
+ * @param {Object} event
+ * @returns {Object} NO_OPERATION_SUCCESS
+ */
 const processIdentifierEvent = async (event) => {
-  const { cartToken, anonymousId, userId } = event;
-  if (cartToken && anonymousId) {
+  const { cartToken, anonymousId, userId, action } = event;
+  if (cartToken && anonymousId && action === 'stitchCartTokenToAnonId') {
```
Comment (utsabc) [inline src/v1/sources/shopify/utils.js:32]: Add stats for both based on actions

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/457/index.html

File: src/v1/sources/shopify/utils.js (hunk 11 above)
Comment (yashasvibajpai) [reply src/v1/sources/shopify/utils.js:38]: Added

**Review by utsabc — DISMISSED** [review]

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/458/index.html

**Review by krishna2020 — DISMISSED** [review]

**Comment by koladilip** [comment]
> ## [Codecov](https://app.codecov.io/gh/rudderlabs/rudder-transformer/pull/4060?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs) Report
> 
> Attention: Patch coverage is `94.44444%` with `3 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 91.13%. Comparing base [(`69a1d6b`)](https://app.codecov.io/gh/rudderlabs/rudder-transformer/commit/69a1d6b7c72847cbb2c00245e60678c1b8152800?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs) to head [(`e35a628`)](https://app.codecov.io/gh/rudderlabs/rudder-transformer/commit/e35a6285e20559ca530f3aa7c92c00b2bed2e525?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs).
> > Report is 38 commits behind head on develop.
> 
> [Files with missing lines](https://app.codecov.io/gh/rudderlabs/rudder-transformer/pull/4060?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs) 	Patch % 	Lines
> [.../shopify/webpixelTransformations/pixelTransform.js](https://app.codecov.io/gh/rudderlabs/rudder-transformer/pull/4060?src=pr&el=tree&filepath=src%2Fv1%2Fsources%2Fshopify%2FwebpixelTransformations%2FpixelTransform.js&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs#diff-c3JjL3YxL3NvdXJjZXMvc2hvcGlmeS93ZWJwaXhlbFRyYW5zZm9ybWF0aW9ucy9waXhlbFRyYW5zZm9ybS5qcw==) 	77.77% 	[2 Missing ⚠️ ](https://app.codecov.io/gh/rudderlabs/rudder-transformer/pull/4060?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs)
> [src/v1/sources/shopify/utils.js](https://app.codecov.io/gh/rudderlabs/rudder-transformer/pull/4060?src=pr&el=tree&filepath=src%2Fv1%2Fsources%2Fshopify%2Futils.js&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs#diff-c3JjL3YxL3NvdXJjZXMvc2hvcGlmeS91dGlscy5qcw==) 	90.90% 	[1 Missing ⚠️ ](https://app.codecov.io/gh/rudderlabs/rudder-transformer/pull/4060?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs)
> Additional details and impacted files
> 
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/rudderlabs/rudder-transformer/pull/4060?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rudderlabs).

@yashasvibajpai fix these

**Comment by koladilip** [comment]
> ## [Quality Gate Passed](https://sonarcloud.io/dashboard?id=rudderlabs_rudder-transformer&pullRequest=4060) **Quality Gate passed**
> 
> Issues  [0 New issues](https://sonarcloud.io/project/issues?id=rudderlabs_rudder-transformer&pullRequest=4060&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)  [0 Accepted issues](https://sonarcloud.io/project/issues?id=rudderlabs_rudder-transformer&pullRequest=4060&issueStatuses=ACCEPTED)
> 
> Measures  [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=rudderlabs_rudder-transformer&pullRequest=4060&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)  [92.3% Coverage on New Code](https://sonarcloud.io/component_measures?id=rudderlabs_rudder-transformer&pullRequest=4060&metric=new_coverage&view=list)  [9.3% Duplication on New Code](https://sonarcloud.io/component_measures?id=rudderlabs_rudder-transformer&pullRequest=4060&metric=new_duplicated_lines_density&view=list)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=rudderlabs_rudder-transformer&pullRequest=4060)

Fix sonar duplication issues

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: https://rudderlabs.github.io/rudder-transformer/462/index.html

**Review by utsabc — APPROVED** [review]

**Review by krishna2020 — APPROVED** [review]
